### ISSUE TITLE
upgrade to terra

### ISSUE THREAD

**Issue by DanWismer** [NCC-CNC/wtw-data-prep#25]

**Comment by DanWismer** [comment]
@edwardsmarc I have updated to using `terra` on branch #24-renaming:
https://github.com/NCC-CNC/wtw-data-prep/blob/%2324-renaming/scripts/03_natdata_to_1km_pu_grid.R

The new formatted data I put on the R drive here: **R:\NCCGIS\PROJECTS\NATGIS\CONS_TECH\PRZ_TOOL\DATA_PREP\WTW_NAT_DATA_20230706.zip**

After getting it to run on the `PU.tif` (attached), I am noticing a slower run time AND the use of more RAM ...

[PU.zip](https://github.com/NCC-CNC/wtw-data-prep/files/11983673/PU.zip)

When you have time, could you give this a run and see if your experiencing the same? I am wondering if there are some settings in `terra` that need to be specified to make things more optimized?

**Comment by DanWismer** [comment]
... updated terra branch with that SW-Ontario PU.tif took 7 mins :( ... with main branch using raster, I have it at ~4 mins

**Comment by edwardsmarc** [comment]
Interesting! I'll test this on Monday. Maybe I can do some benchmarking to figure out which functions are causing the slow down. I'd guess its the `terra::crop` since most the other stuff is just converting formats and saving.

**Comment by DanWismer** [comment]
update: **WTW_NAT_DATA_20230710.zip**

**Comment by edwardsmarc** [comment]
The slow down is occurring in the function matrix_to_raster.

terra::crop() actually appears to be faster than raster::crop(), but overwriting the raster values in a terra object is very slow. These two lines are responsible for all the slow down as far as I can tell:

```
ncc_1km_idx[] <- NA
ncc_1km_idx[natdata_intersect["Idx",]] <- natdata_intersect[i,]
```

Some speed can be gained by taking `ncc_1km_idx[] <- NA` out of the for loop, but the remaining line is still very slow.

I'll look into other options for building the raster.

**Comment by DanWismer** [comment]
I found terra::crop pretty fast as well.

Nice catch, I am wondering if instead of building the raster inside the terra object, we build it as a `df` first, then convert it to terra?

**Comment by edwardsmarc** [comment]
@DanWismer I just pushed some changes that speed up the script. I had to merge your edits from yesterday back in, hopefully I didn't miss any.

Turns out terra is just bad at setting values in an existing spatRaster object. They even [recommend avoiding](https://rspatial.github.io/terra/reference/setValues.html) this for large rasters and using terra::init() instead.

I switched to init() for initializing the PU raster with NA values. I also moved this out of fct_raster_to_matrix() so it only has to be done once in the script.

Our problem is that we have to update the values of the raster using a subset. Because we initialize a raster with NAs, then fill in values overlapping the AOI by grabbing the correctly indexed values from our rij matrix. As far as I can tell there is no way to do this using the init() function. We're stuck with setValues.

See how this fix works. For me the old raster package script takes 2.14 mins and this one takes 2.40 mins.

We might be able to speed it up further by avoiding setValues completely. Maybe by converting the AOI and Theme data to a spatRaster and using or ifel() to set the NA values.

Note: I did try your suggestion of setting the NA values to non-AOI PUs in the matrix, then converting the matrix rows straight to spatRasters. Setting the NA values took way to long which makes sense based on the number of PUs.
